src/char/int_achievement.c

```diff
@@ -0,0 +1,341 @@
+
+/**
+ * Saves changed achievements for a character.
+ * @param  char_id     character identifier.
+ * @param  cp          pointer to loaded achievements.
+ * @param  p           pointer to map-sent character achievements.
+ * @return number of achievements saved.
+ */
+int inter_achievement_tosql(int char_id, struct char_achievements *cp, struct char_achievements *p)
+{
+	StringBuf buf;
+	int i = 0, rows = 0;
+
+	nullpo_ret(cp);
+	nullpo_ret(p);
+	Assert_ret(char_id > 0);
+
+	StrBuf->Init(&buf);
+	StrBuf->Printf(&buf, "REPLACE INTO `%s` (`char_id`, `ach_id`, `completed_at`, `rewarded_at`", char_achievement_db);
+	for (i = 0; i < MAX_ACHIEVEMENT_OBJECTIVES; i++)
+		StrBuf->Printf(&buf, ", `obj_%d`", i);
+	StrBuf->AppendStr(&buf, ") VALUES ");
+
+	for (i = 0; i < VECTOR_LENGTH(*p); i++) {
+		int j = 0;
+		bool save = false;
+		struct achievement *pa = &VECTOR_INDEX(*p, i), *cpa = NULL;
+
+		ARR_FIND(0, VECTOR_LENGTH(*cp), j, ((cpa = &VECTOR_INDEX(*cp, j)) && cpa->id == pa->id));
+
+		if (j == VECTOR_LENGTH(*cp))
+			save = true;
+		else if (memcmp(cpa, pa, sizeof(struct achievement)) != 0)
+			save = true;
+
+		if (save) {
+			StrBuf->Printf(&buf, "%s('%d', '%d', '%"PRId64"', '%"PRId64"'", rows ?", ":"", char_id, pa->id, (int64)pa->completed_at, (int64)pa->rewarded_at);
+			for (j = 0; j < MAX_ACHIEVEMENT_OBJECTIVES; j++)
+				StrBuf->Printf(&buf, ", '%d'", pa->objective[j]);
+			StrBuf->AppendStr(&buf, ")");
+			rows++;
+		}
+	}
+
+	if (rows > 0 && SQL_ERROR == SQL->QueryStr(inter->sql_handle, StrBuf->Value(&buf))) {
+		Sql_ShowDebug(inter->sql_handle);
+		StrBuf->Destroy(&buf); // Destroy the buffer.
+		return 0;
+	}
+	// Destroy the buffer.
+	StrBuf->Destroy(&buf);
+
+	if (rows) {
+		ShowInfo("achievements saved for char %d (total: %d, saved: %d)\n", char_id, VECTOR_LENGTH(*p), rows);
+
+		/* Sync with inter-db acheivements. */
+		VECTOR_CLEAR(*cp);
+		*cp = *p;
+	}
+
+	return rows;
+}
+
+/**
+ * Retrieves all achievements of a character.
+ * @param char_id  character identifier.
+ * @param ca       pointer to character achievements structure.
+ * @return true on success, false on failure.
+ */
+bool inter_achievement_fromsql(int char_id, struct char_achievements *cp)
+{
+	StringBuf buf;
+	char *data;
+	int i = 0, num_rows = 0;
+
+	Assert_ret(char_id > 0);
+	nullpo_ret(cp);
+
+	// char_achievements (`char_id`, `ach_id`, `completed_at`, `rewarded_at`, `obj_0`, `obj_2`, ...`obj_9`)
+	StrBuf->Init(&buf);
+	StrBuf->AppendStr(&buf, "SELECT `ach_id`, `completed_at`, `rewarded_at`");
+	for (i = 0; i < MAX_ACHIEVEMENT_OBJECTIVES; i++)
+		StrBuf->Printf(&buf, ", `obj_%d`", i);
+	StrBuf->Printf(&buf, " FROM `%s` WHERE `char_id` = '%d' ORDER BY `ach_id`", char_achievement_db, char_id);
+
+	if (SQL_ERROR == SQL->QueryStr(inter->sql_handle, StrBuf->Value(&buf))) {
+		Sql_ShowDebug(inter->sql_handle);
+		StrBuf->Destroy(&buf);
+		return false;
+	}
+
+	if ((num_rows = (int) SQL->NumRows(inter->sql_handle)) != 0) {
+		int j = 0;
+
+		VECTOR_INIT(*cp);
```

Comment:
This is actually initialised in db ensuring function so I'll remove it from here as it is unnecessary.